experiments/process_sandbox/include/process_sandbox/platform/sandbox_seccomp-bpf.h

```diff
@@ -2,12 +2,12 @@
 // SPDX-License-Identifier: MIT
 
 #pragma once
-#if __has_include(<seccomp.h>)
-#  pragma clang diagnostic push
-#  pragma clang diagnostic ignored "-Wmissing-field-initializers"
+#include "../helpers.h"
 
+#if __has_include(<seccomp.h>)
+SANDBOX_CLANG_DIAGNOSTIC_IGNORE("-Wmissing-field-initializers")
 #  include <seccomp.h>
-#  pragma clang diagnostic pop
+SANDBOX_CLANG_DIAGNOSTIC_POP()
```

Comment:
I don't want to disable this warning globally, only for this header (which is a Linux system header and so I can't fix it).  C doesn't let you have different warning flags for different directories: warnings can be disabled by compiler flags only per compilation unit, so disabling it for this header would require disabling it for everything that includes this header.

These headers are part of the installed set and so can be used by other things, so forcing consumers to disable a (potentially useful) warning seems rude.

Reply:
I get it why you want this and why it's fine like you did, I'm not challenging that. I just think this is an overall bad pattern.

If this is going to be a wider pattern in Verona, I'd like it be consistent. There are other cases that we have disabled warnings that may want to be converted to the same method, but also, I don't think this is a clang-specific behaviour, or worse, it may have different behaviour in different compilers, including gcc.

I'm not sure what's the feature matrix between warnings, but the more skips we add like that, the more likely a warning won't be available on a clang of an old enough version and give errors. If the decision is in CMake, the fix is easier than if the decision is peppered all over the source code.

Sure, adding warnings in CMake per directory (ex. `src/interop`) may hide some useful warnings, but it's a lot more maintainable. What happens when headers change, on different systems, when the underlying code changes so the warning won't happen anymore? How do we know it's now safe to remove the diagnostics macro?

Reply:
We are already using this model in snmalloc [here](https://github.com/microsoft/snmalloc/blob/b598add0d1adc9b8ad311b0928f63254fe69073c/src/mem/threadalloc.h#L62) for MSVC.

> I don't think this is a clang-specific behaviour, or worse, it may have different behaviour in different compilers, including gcc.

It is explicitly for things that are different between different compilers.  Clang, gcc, and MSVC all have equivalent pragmas for pushing warnings state, disabling specific warnings, and popping warning state.  

> I'm not sure what's the feature matrix between warnings, but the more skips we add like that, the more likely a warning won't be available on a clang of an old enough version and give errors.

I'm not sure what this means.  

> If the decision is in CMake, the fix is easier than if the decision is peppered all over the source code.

If the decision is in CMake then you're disabling the warning for entire compilation units.  In a header-only or mostly header project, per compilation is equivalent to global.  If a warnings gives mostly false positives, we should disable it.  If it's mostly fine but has false positives in a few places, we should silence them.  Sometimes we can silence them by modifying the code (e.g. by doing a `void` cast to silence unused variable warnings or by adding assert or assume builtins).  This mechanism is for where we can't do that.

> Sure, adding warnings in CMake per directory (ex. src/interop) may hide some useful warnings, but it's a lot more maintainable

I don't think it's that clear cut.  Silencing warnings increases the probability of bugs surviving CI, so it's a problem for maintenance.  Silencing warnings globally also has most of the same problems you outline:

> What happens when headers change, on different systems, when the underlying code changes so the warning won't happen anymore? How do we know it's now safe to remove the diagnostics macro?

What happens in all of these situations with the warning disabled in CMake?  Exactly the same thing: We have to try, test, and see.  If the warnings are disabled for a tiny scope, there's a much lower impact to not bothering to remove them.  In this case, for example, warning is not disabled for *any* of our code and so if the system header is fixed then that's great but it doesn't cost us anything to leave the warning disabled when we include it.  If we had disabled the warning globally, then it is costing us some useful checking in CI (and on local builds) and so we want to remove the CMake logic disabling it as soon as possible.